shaman.1
'\" t
.\" Manual page created with latex2man on Fri Nov 29 20:38:57 CST 2013
.\" NOTE: This file is generated, DO NOT EDIT.
.de Vb
.ft CW
.nf
..
.de Ve
.ft R

.fi
..
.TH "SHAMAN" "1" "29 November 2013" "weather fetcher " "weather fetcher "
.SH NAME

.PP
Shaman
\- Fetch basic weather information from the National Weather Service 
.PP
.SH SYNOPSIS

shaman
[\fB\-h\fP]
[\fB\-a\fP]
[\fB\-i|\-m\fP]
[\fB\-c\fP]
[\fB\-d\fP]
[\fB\-D\fP]
[\fB\-l\fP]
[\fB\-p\fP]
[\fB\-r\fP]
[\fB\-t\fP]
[\fB\-v\fP]
[\fB\-w\fP]
[\fB\-z\fP]
\fI"location"\fP
.PP
.SH DESCRIPTION

Shaman
is a very lightweight application to fetch weather information from the National Weather Service using \fIcurl\fP(1)\&.
.PP
.SH OPTIONS

.TP
\fB\-h\fP
 Print help and exit 
.TP
\fB\-a\fP
 Print all available information 
(Assumed if no options are passed) 
.TP
\fB\-i\fP, \fB\-m\fP
 Print information using Imperial or Metric units (respectively) 
.TP
\fB\-c\fP
 Print current weather conditions 
.TP
\fB\-d\fP
 Print humidity 
.TP
\fB\-D\fP
 Print dew\-point 
.TP
\fB\-l\fP
 Print heat index 
.TP
\fB\-p\fP
 Print pressure 
.TP
\fB\-r\fP
 Print reporter information 
.TP
\fB\-t\fP
 Print temperature 
.TP
\fB\-v\fP
 Print visibility 
.TP
\fB\-w\fP
 Print wind speed and direction 
.TP
\fB\-z\fP
 Suppress printing of hazardous and severe weather warnings 
.TP
\fI"location"\fP
 Zip code or city name (formatted as "City, ST") of the location to fetch information for 
.PP
.SH CONFIGURATION

A very minimal configuration file may be used to tell shaman
some default values: a default location, and whether to use Imperial or Metric units by default. 
Shaman
will look for this config file in the following places in the following order, and will stop searching and load the first file it finds: 
$XDG_CONFIG_HOME/shaman/config,
$HOME/.config/shaman/config,
$HOME/.shamanrc
.PP
The config file should be formatted in the following way: 
.PP
Location=\fI"location"\fP
.br
Units="U" 
.PP
As with its command\-line argument counter\-part, \fI"location"\fP
may be formatted as either the zipcode or name. 
"U" may be either an "I" or an "M" (for Imperial and Metric respectively). 
.PP
.SH AUTHOR

Copyright (C)2012\-2013 Sam Stuewe 
.br
License GPLv2: GNU GPL version 2 \fBhttps://gnu.org/licenses/old\-licenses/gpl\-2.0.html\fP
.br
This is free software: you are free to change and redistribute it. 
.br
There is NO WARRANTY, to the extent permitted by law. 
.PP
Bug reports, suggestions and comments may be submitted via GitHub (\fBhttp://github.com/HalosGhost/shaman.git\fP)
or email (\fBhalosghost@archlinux.info\fP).
.PP
.SH SEE ALSO

\fIcurl\fP(1)
.PP
.\" NOTE: This file is generated, DO NOT EDIT.


shaman.c
/****************************************\
* Fetch weather information from the NWS *
* Author: 2012-2013 (C) Sam Stuewe       *
* License: GPLv2                         *
\****************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>
#include <math.h>
#include <curl/curl.h>
#include <libconfig.h>
#include <getopt.h>

// Variables //
FILE *url;
static int flag_help, flag_metric;
int defaultUnits;
char formatString[80],
	 passLoc[80];

// Prototypes //
void usage(char *progname);
void discoverConfig(void);
void parseConfig(char *configfile);
void getData(char *location);
void getReporter(char *reporterln, char *coordsln, char *elevln);
void getConditions(char *conditionsln);
void getHazards(char *hazardln);
void parseFormat(char *formatStr);

// Main Function //
int main(int argc, char **argv)
{   if ( argc > 1 )
	{   int c;

		while (1) 
		{	static struct option long_options[] = 
			{   // Long Option Flags //
				{"help",	  no_argument,		   0,	 'h'   },
				{"imperial",  no_argument,		   0,	 'i'   },
				{"metric",	  no_argument,		   0,	 'm'   },
				// Long Option Switches //
				{"config",	  required_argument,   0,	 'c'   },
				{"format",	  required_argument,   0,	 'f'   },
				{"location",  required_argument,   0,	 'l'   },
				{0,			  0,				   0,	 0	   },
			};

			int option_index = 0; // Stores the option index

			c = getopt_long(argc, argv, "hc:if:l:m", long_options, &option_index);

			if ( c == -1 ) break;

			switch (c)
			{   case 'h':
				    flag_help = 1;
					break;
				
				case 'c':
					parseConfig(optarg);
					break;

			    case 'i':
					flag_metric = 0;
					break;

			    case 'f':
					strncpy(formatString, optarg, strlen(formatString));
					break;

			    case 'l':
					strncpy(passLoc, optarg, strlen(passLoc));
					break;

			    case 'm':
					flag_metric = 1;
					break;
			}
		}
	}

	if ( flag_help == 1 ) usage(argv[0]);
	return 0;
}

// Usage Function //
void usage(char *progname)
{   fprintf(stderr, "Usage: %s [-h] [-c configfile] [-i|-m] [-f \"<format string>\"] -l \"<location>\"\n\n", progname);
	fprintf(stderr, "Options:\n");
	fprintf(stderr, "  -h, --help                    print help and exit\n");
	fprintf(stderr, "  -c, --config=filename         use specified config file\n");
	fprintf(stderr, "  -i, --imperial                use Imperial units (default)\n");
	fprintf(stderr, "  -m, --metric                  use Metric units\n");
	fprintf(stderr, "  -f, --format=\"formatString\"   format output according to \"formatString\"\n");
	fprintf(stderr, "  -l, --location=\"location\"     print weather information for \"location\"\n");
	fprintf(stderr, "\nSee `man 1 %s` for more information\n", progname);
	exit(0);
}

/* Configuration Functions //
void discoverConfig(void)
{   config_t(cfg);
	const config_setting_t *Location, *Format, defaultUnits;
	const char *cwd = getenv("PWD");
	char *confString = "config";

	if ( !*passLoc && !*formatString && !defaultUnits )
    {   if ( chdir(getenv("XDG_CONFIG_HOME")) )
		{	if ( chdir("shaman") ) 
			{   if ( access("config", R_OK) != -1 ) config_init("config");
			}
		}
		else 
	    {	chdir(getenv("HOME"));
			if ( chdir(".config/shaman") ) access("config", R_OK);
			else 
			{   access(".shamanrc", R_OK);
				confString = ".shamanrc";
			}
		}
		config_lookup_string("Location", confString, passLoc);
		config_lookup_string("Format", confString, formatString);
		config_lookup_bool("Use_Metric", confString, flag_metric);
	}
	chdir(cwd);
}*/

void parseConfig(char *configFile)
{   printf("Function not yet implemented\n");
}

// Data and Analysis Functions //
void getData(char *location)
{   printf("Function not yet implemented\n");
}

void getReporter(char *reporterln, char *coordsln, char *elevln)
{   printf("Function not yet implemented\n");
}

void getConditions(char *conditionsln)
{   printf("Function not yet implemented\n");
}

void getHazards(char *hazardln)
{   printf("Function not yet implemented\n");
}

void parseFormat(char *formatStr)
{   printf("Function not yet implemented\n");
}
